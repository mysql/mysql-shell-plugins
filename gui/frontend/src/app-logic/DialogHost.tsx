/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import {
    ConfirmDialog, DialogValueOption, IDialogSection, IDialogValues, PasswordDialog, ValueDialogBase, ValueEditDialog,
} from "../components/Dialogs";
import { Component } from "../components/ui";
import { MrsSchemaDialog } from "../modules/mrs/dialogs/MrsSchemaDialog";

import { MrsServiceDialog } from "../modules/mrs/dialogs/MrsServiceDialog";
import { requisitions } from "../supplement/Requisitions";
import { DialogResponseClosure, DialogType, IDialogRequest, IDialogResponse, IDictionary } from "./Types";

/**
 * A component to host certain application wide dialogs in a central place.
 * They are all accessible via requisitions.
 */
export class DialogHost extends Component {
    // Holds the currently running dialog type (only one of each type can run at the same time) and last
    // active HTML element, when this dialog was launched.
    private runningDialogs = new Map<DialogType, Element | null>();

    private dialogRefs = new Map<DialogType, React.RefObject<ValueDialogBase>>();

    // The value edit dialog is special, as it uses a different approach for editing complex values.
    private promptDialogRef = React.createRef<ValueEditDialog>();

    private confirmDialogRef = React.createRef<ConfirmDialog>();

    public constructor(props: {}) {
        super(props);

        requisitions.register("showDialog", this.showDialog);
    }

    public render(): React.ReactNode {
        const dialogs: React.ReactNode[] = [
            // The password dialog has it's own command handling. Just host it here.
            <PasswordDialog key="passwordDialog" />,

            // The value edit dialog has a different value handling, so it's not added to the dialogRefs list.
            <ValueEditDialog
                key="valueEditDialog"
                ref={this.promptDialogRef}
                onClose={this.handlePromptDialogClose}
            />,

            <ConfirmDialog
                key="confirmDialog"
                ref={this.confirmDialogRef}
                onClose={this.handleDialogClose.bind(this, DialogType.Confirm)}
            />,
        ];

        const ref1 = React.createRef<MrsServiceDialog>();
        this.dialogRefs.set(DialogType.MrsService, ref1);
        dialogs.push(<MrsServiceDialog
            key="mrsServiceDialog"
            ref={ref1}
            onClose={this.handleDialogClose.bind(this, DialogType.MrsService)}
        />);

        const ref2 = React.createRef<MrsSchemaDialog>();
        this.dialogRefs.set(DialogType.MrsSchema, ref2);
        dialogs.push(<MrsSchemaDialog
            ref={ref2}
            onClose={this.handleDialogClose.bind(this, DialogType.MrsSchema)}
        />);

        return (
            <>
                {dialogs}
            </>
        );
    }

    private showDialog = (request: IDialogRequest): Promise<boolean> => {
        // Check if a dialog of the given type is already active.
        // Only one of each type can be active at any time.
        if (!this.runningDialogs.has(request.type)) {
            switch (request.type) {
                case DialogType.Prompt: {
                    this.runPromptDialog(request);

                    return Promise.resolve(true);
                }

                case DialogType.Confirm: {
                    this.runConfirmDialog(request);

                    return Promise.resolve(true);
                }

                case DialogType.Select: {
                    this.runSelectDialog(request);

                    return Promise.resolve(true);
                }

                default: { // All dialogs with the base value editor return signature.
                    const ref = this.dialogRefs.get(request.type);
                    if (ref && ref.current) {
                        this.runningDialogs.set(request.type, document.activeElement);
                        ref.current.show(request, request.title);

                        return Promise.resolve(true);
                    }

                    break;
                }
            }
        }

        return Promise.resolve(false);
    };

    /**
     * Configures and runs a value edit dialog with a single section to let the user input a single value.
     *
     * Supported entries in the request are:
     *   - request.title The dialog's title.
     *   - request.values.prompt A caption for the input field.
     *   - request.data: A dictionary that is forwarded to the response handler.
     *
     * @param request The request with the data for the dialog.
     */
    private runPromptDialog = (request: IDialogRequest): void => {
        this.runningDialogs.set(DialogType.Prompt, document.activeElement);

        const promptSection: IDialogSection = {
            values: {
                input: {
                    caption: request.values?.prompt as string,
                    value: "",
                    span: 8,
                    options: [DialogValueOption.AutoFocus],
                },
            },
        };

        this.promptDialogRef.current?.show(
            {
                id: request.id,
                sections: new Map<string, IDialogSection>([
                    ["prompt", promptSection],
                ]),
            },
            {
                options: { backgroundOpacity: 0.5 },
                title: request.title ?? "Feedback Requested",
                description: request.description,
            },
            { id: request.id, ...request.data, type: request.type },
        );
    };

    /**
     * Configures and runs a confirmation dialog.
     *
     * Supported entries in the request are:
     *   - parameters.title The dialog's title.
     *   - parameters.prompt The text to show for the confirmation.
     *   - parameters.accept Optional text for the accept button (default: "OK").
     *   - parameters.refuse Optional text for the refuse button (default: "Cancel").
     *   - parameters.alternative Optional text for the accept button (no default).
     *   - parameters.default Optional text for the button that should be auto focused.
     *   - request.data: A dictionary that is forwarded to the response handler.
     *
     * @param request The request with the data for the dialog.
     */
    private runConfirmDialog = (request: IDialogRequest): void => {
        this.runningDialogs.set(DialogType.Confirm, document.activeElement);

        this.confirmDialogRef.current?.show(
            request.parameters?.prompt as string ?? "",
            {
                accept: request.parameters?.accept as string ?? "",
                refuse: request.parameters?.refuse as string ?? "",
                alternative: request.parameters?.alternative as string,
                default: request.parameters?.default as string,
            },
            request.parameters?.title as string,
            request.description,
            { id: request.id, ...request.data, type: request.type },
        );
    };

    /**
     * Configures and runs a selection dialog.
     *
     * Supported entries in the request are:
     *   - parameters.title The dialog's title.
     *   - parameters.prompt The text to show for the confirmation.
     *   - parameters.default Optional text for the button that should be auto focused.
     *   - parameters.options The list of values from which one must be selected.
     *   - request.data: A dictionary that is forwarded to the response handler.
     *
     * @param request The request with the data for the dialog.
     */
    private runSelectDialog = (request: IDialogRequest): void => {
        this.runningDialogs.set(DialogType.Select, document.activeElement);

        const promptSection: IDialogSection = {
            values: {},
        };

        request.description?.forEach((entry, index) => {
            promptSection.values[`description${index}`] = {
                value: entry,
                span: 8,
                options: [DialogValueOption.Description],
            };
        });

        const choices = request.parameters?.options as string[];
        const defaultValue: number | undefined = request.parameters?.default as number;
        promptSection.values.input = {
            caption: request.parameters?.prompt as string,
            value: defaultValue === undefined ? "" : choices[defaultValue - 1], // One-based value.
            span: 8,
            choices,
            options: [DialogValueOption.AutoFocus],
        };

        this.promptDialogRef.current?.show(
            {
                id: request.id,
                sections: new Map<string, IDialogSection>([
                    ["prompt", promptSection],
                ]),
            },
            {
                options: { backgroundOpacity: 0.5 },
                title: request.title ?? "Feedback Requested",
            },
            { id: request.id, ...request.data, type: request.type },
        );
    };

    private handleDialogClose = (type: DialogType, closure: DialogResponseClosure, data?: IDictionary): void => {
        const element = this.runningDialogs.get(type);
        this.runningDialogs.delete(type);

        const response: IDialogResponse = {
            id: data?.id as string ?? "",
            type,
            closure,
            data,
        };

        void requisitions.execute("dialogResponse", response);

        if (element instanceof HTMLTextAreaElement || element instanceof HTMLInputElement) {
            element.focus();
        }
    };

    private handlePromptDialogClose = (closure: DialogResponseClosure, values: IDialogValues,
        data?: IDictionary): void => {

        const type = data?.type as DialogType ?? DialogType.Prompt;
        const element = this.runningDialogs.get(type);
        this.runningDialogs.delete(type);

        const promptSection = values.sections.get("prompt");
        if (promptSection) {
            let text = promptSection.values.input.value as string;
            if (type === DialogType.Select) {
                // Convert the text to an index in the choice list.
                const index = promptSection.values.input.choices?.findIndex((value) => {
                    return value === text;
                }) ?? -1;

                if (index > -1) {
                    text = String(index + 1);
                }
            }

            const response: IDialogResponse = {
                type: data?.type as DialogType,
                id: data?.id as string ?? "",
                closure,
                values: {
                    input: text,
                },
                data,
            };

            void requisitions.execute("dialogResponse", response);
        }

        if (element instanceof HTMLTextAreaElement || element instanceof HTMLInputElement) {
            element.focus();
        }
    };

}
