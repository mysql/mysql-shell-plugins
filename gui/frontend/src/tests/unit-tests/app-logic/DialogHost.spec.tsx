/*
 * Copyright (c) 2021, 2023, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import keyboardKey from "keyboard-key";

import { mount } from "enzyme";

import { DialogHost } from "./../../../app-logic/DialogHost";
import { JestReactWrapper, nextProcessTick, sendKeyPress } from "../test-helpers";
import { DialogType, IDialogRequest } from "../../../app-logic/Types";
import { requisitions } from "../../../supplement/Requisitions";

describe("DialogHost Tests", () => {
    let host: JestReactWrapper;

    beforeAll(() => {
        host = mount<DialogHost>(<DialogHost />);
    });

    afterAll(() => {
        host.unmount();
    });

    it("Standard Rendering (snapshot)", () => {
        // The host itself has no properties, but implicit children (the different dialogs).
        expect(host.props().children).toEqual([]);
        expect(host).toMatchSnapshot();
    });

    it("Show Prompt Dialog (snapshot)", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);

        const promptRequest: IDialogRequest = {
            type: DialogType.Prompt,
            id: "shellPromptDialog",
            values: {
                prompt: "Lorem ipsum dolor sit amet",
            },
            data: {
                requestId: "12345",
            },
        };

        await requisitions.execute("showDialog", promptRequest);
        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(1);

        expect(portals[0]).toMatchSnapshot();

        // Try to show the dialog again -> should have no effect.
        await requisitions.execute("showDialog", promptRequest);
        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(1);

        sendKeyPress(keyboardKey.Escape);
        await nextProcessTick();

        portals = document.getElementsByClassName("portal");
        expect(portals.length).toBe(0);
    });

});
