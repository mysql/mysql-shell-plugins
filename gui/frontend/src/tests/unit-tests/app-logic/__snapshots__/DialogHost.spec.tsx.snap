// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogHost Tests Show MRS Schema Dialog (snapshot) 1`] = `
<div
  __self="[object Object]"
  __source="[object Object]"
  class="msg portal"
  id="mrsSchemaDialog"
>
  <div
    __self="[object Object]"
    __source="[object Object]"
    class="msg dialog visible valueEditDialog"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="title verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg icon codicon codicon-terminal"
      />
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip="expand"
      >
        MySQL REST Service
      </label>
      <button
        __self="[object Object]"
        __source="[object Object]"
        class="msg button imageOnly"
        id="closeButton"
        tabindex="0"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg icon"
          style="mask-image: url(close2.svg);"
        />
      </button>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="content fixedScrollbar"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg grid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Schema Name
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="name"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            MRS Service
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar dropdown valueEditor"
            id="service"
            style="flex-direction: row; align-items: center;"
            tabindex="0"
          >
            <label
              __self="[object Object]"
              __source="[object Object]"
              class="msg label ellipsis"
              data-tooltip="expand"
            >
              localhost
            </label>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Request Path
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="requestPath"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Requires Authentication
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="requiresAuth"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Items per Page
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="itemsPerPage"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Enabled
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="enabled"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Comments
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="comments"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="footer verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="OK"
          class="msg button"
          id="ok"
          tabindex="0"
        >
          OK
        </button>
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="Cancel"
          class="msg button"
          id="cancel"
          tabindex="0"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DialogHost Tests Show MRS Service Dialog (snapshot) 1`] = `
<div
  __self="[object Object]"
  __source="[object Object]"
  class="msg portal"
  id="mrsServiceDialog"
>
  <div
    __self="[object Object]"
    __source="[object Object]"
    class="msg dialog visible valueEditDialog"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="title verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg icon codicon codicon-terminal"
      />
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip="expand"
      >
        MySQL REST Service
      </label>
      <button
        __self="[object Object]"
        __source="[object Object]"
        class="msg button imageOnly"
        id="closeButton"
        tabindex="0"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg icon"
          style="mask-image: url(close2.svg);"
        />
      </button>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="content fixedScrollbar"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg grid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Service Name
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="serviceName"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Comments
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="comments"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Host Name
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="hostName"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Supported Protocols
          </label>
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label description"
            data-tooltip="expand"
          >
            Supported Protocols
          </label>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            MRS Service Flags
          </label>
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg checkbox unchecked valueEditor"
            id="makeDefault"
            tabindex="0"
          >
            <span
              __self="[object Object]"
              __source="[object Object]"
              class="checkMark"
            />
            Default
          </label>
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg checkbox checked valueEditor"
            id="enabled"
            tabindex="0"
          >
            <span
              __self="[object Object]"
              __source="[object Object]"
              class="checkMark"
            />
            Enabled
          </label>
        </div>
      </div>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="footer verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="OK"
          class="msg button"
          id="ok"
          tabindex="0"
        >
          OK
        </button>
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="Cancel"
          class="msg button"
          id="cancel"
          tabindex="0"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DialogHost Tests Show Prompt Dialog (snapshot) 1`] = `
<div
  __self="[object Object]"
  __source="[object Object]"
  class="msg portal"
  id="shellPromptDialog"
>
  <div
    __self="[object Object]"
    __source="[object Object]"
    class="msg dialog visible valueEditDialog"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="title verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg icon codicon codicon-terminal"
      />
      <label
        __self="[object Object]"
        __source="[object Object]"
        class="msg label"
        data-tooltip="expand"
      >
        Feedback Requested
      </label>
      <button
        __self="[object Object]"
        __source="[object Object]"
        class="msg button imageOnly"
        id="closeButton"
        tabindex="0"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg icon"
          style="mask-image: url(close2.svg);"
        />
      </button>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="content fixedScrollbar"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg grid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
        >
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Lorem ipsum dolor sit amet
          </label>
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              __self="[object Object]"
              __source="[object Object]"
              class="msg input valueEditor"
              id="input"
              spellcheck="true"
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="footer verticalCenterContent"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="OK"
          class="msg button"
          id="ok"
          tabindex="0"
        >
          OK
        </button>
        <button
          __self="[object Object]"
          __source="[object Object]"
          caption="Cancel"
          class="msg button"
          id="cancel"
          tabindex="0"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DialogHost Tests Standard Rendering (snapshot) 1`] = `
<DialogHost>
  <PasswordDialog
    key="passwordDialog"
  />
  <ValueEditDialog
    key="valueEditDialog"
    onClose={[Function]}
  >
    <Dialog
      actions={
        Object {
          "begin": Array [],
          "end": Array [
            <Button
              caption="OK"
              disabled={false}
              id="ok"
              onClick={[Function]}
            />,
            <Button
              caption="Cancel"
              id="cancel"
              onClick={[Function]}
            />,
          ],
        }
      }
      caption={
        <y>
          <Icon
            disabled={false}
            src={388}
          />
          <Label />
        </y>
      }
      className="msg valueEditDialog"
      container={Object {}}
      content={
        <Grid
          columnGap={16}
          columns={8}
          rowGap={16}
        />
      }
      id=""
      onClose={[Function]}
    >
      <Portal
        container={Object {}}
        id=""
        onClose={[Function]}
        onOpen={[Function]}
      />
    </Dialog>
  </ValueEditDialog>
  <ConfirmDialog
    key="confirmDialog"
    onClose={[Function]}
  >
    <Dialog
      actions={
        Object {
          "end": Array [],
        }
      }
      caption={
        <y>
          <Icon
            disabled={false}
            src={275}
          />
          <Label>
            Confirm
          </Label>
        </y>
      }
      className="msg confirmDialog"
      container={Object {}}
      content={
        <Container
          orientation="column"
        >
          
          <Container
            className="description"
            orientation="column"
          />
        </Container>
      }
      onClose={[Function]}
    >
      <Portal
        container={Object {}}
        onClose={[Function]}
        onOpen={[Function]}
      />
    </Dialog>
  </ConfirmDialog>
  <MrsServiceDialog
    key="mrsServiceDialog"
    onClose={[Function]}
  >
    <ValueEditDialog
      id="mrsServiceDialog"
      onClose={[Function]}
      onValidate={[Function]}
    >
      <Dialog
        actions={
          Object {
            "begin": Array [],
            "end": Array [
              <Button
                caption="OK"
                disabled={false}
                id="ok"
                onClick={[Function]}
              />,
              <Button
                caption="Cancel"
                id="cancel"
                onClick={[Function]}
              />,
            ],
          }
        }
        caption={
          <y>
            <Icon
              disabled={false}
              src={388}
            />
            <Label />
          </y>
        }
        className="msg valueEditDialog"
        container={Object {}}
        content={
          <Grid
            columnGap={16}
            columns={8}
            rowGap={16}
          />
        }
        id="mrsServiceDialog"
        onClose={[Function]}
      >
        <Portal
          container={Object {}}
          id="mrsServiceDialog"
          onClose={[Function]}
          onOpen={[Function]}
        />
      </Dialog>
    </ValueEditDialog>
  </MrsServiceDialog>
  <MrsSchemaDialog
    onClose={[Function]}
  >
    <ValueEditDialog
      id="mrsSchemaDialog"
      onClose={[Function]}
      onValidate={[Function]}
    >
      <Dialog
        actions={
          Object {
            "begin": Array [],
            "end": Array [
              <Button
                caption="OK"
                disabled={false}
                id="ok"
                onClick={[Function]}
              />,
              <Button
                caption="Cancel"
                id="cancel"
                onClick={[Function]}
              />,
            ],
          }
        }
        caption={
          <y>
            <Icon
              disabled={false}
              src={388}
            />
            <Label />
          </y>
        }
        className="msg valueEditDialog"
        container={Object {}}
        content={
          <Grid
            columnGap={16}
            columns={8}
            rowGap={16}
          />
        }
        id="mrsSchemaDialog"
        onClose={[Function]}
      >
        <Portal
          container={Object {}}
          id="mrsSchemaDialog"
          onClose={[Function]}
          onOpen={[Function]}
        />
      </Dialog>
    </ValueEditDialog>
  </MrsSchemaDialog>
</DialogHost>
`;
